kernel/timer.c

```diff
@@ -32,8 +32,22 @@ void z_timer_expiration_handler(struct _timeout *t)
 	 */
 	if (!K_TIMEOUT_EQ(timer->period, K_NO_WAIT) &&
 	    !K_TIMEOUT_EQ(timer->period, K_FOREVER)) {
+		k_timeout_t next = timer->period;
+
+#ifdef CONFIG_TIMEOUT_64BIT
+		/* Exploit the fact that uptime during a kernel
+		 * timeout handler reflects the time of the scheduled
+		 * event and not real time to get some inexpensive
+		 * protection against late interrupts.  If we're
+		 * delayed for any reason, we still end up calculating
+		 * the next expiration as a regular stride from where
+		 * we "should" have run.  Requires absolute timeouts.
+		 */
+		next = K_TIMEOUT_ABS_TICKS(k_uptime_ticks()
```

Comment:
this is definitely what I'd expect from a periodic timer, and makes good sense I think. It doesn't matter when the interrupt happens, you schedule based on what you wanted in the period not when the event happens